educator/src/Dscp/Educator/Web/Educator/API.hs

```diff
@@ -0,0 +1,230 @@
+        :> Get '[JSON] [SubmissionEducatorInfo]
+
+    , eGetStudentAssignmentSubmissions :: route
+        :- "students"    :> Capture "studentAddr" Student
+        :> "assignments" :> Capture "assignmentHash" (Hash Assignment)
+        :> "submissions"
+        :> Summary "Student's submissions for an assignment"
+        :> Description "Gets a list of student's submissions for a given \
+                    \assignment"
+        :> Get '[JSON] [SubmissionEducatorInfo]
+
+    , eGetStudentCourseSubmissions :: route
+        :- "students" :> Capture "studentAddr" Student
+        :> "courses"  :> Capture "courseId" Course
+        :> "submissions"
+        :> Summary "Get student's course submissions"
+        :> Description "Gets a list of student's submissions he made during \
+                        \studying given course."
+        :> Get '[JSON] [SubmissionEducatorInfo]
+
+      -- * Grades
+
+    , ePostGrade :: route
+        :- "grades"
+        :> Summary "Post a new grade"
+        :> Description "Posts a new grade with a given body."
+        :> ReqBody '[JSON] NewGrade
+        :> PostCreated '[JSON] ()
+
+    , eGetGrades :: route
+        :- "grades"
+        :> Summary "Get all grades"
+        :> Description "Gets a list of all grades performed by all students."
+        :> Get '[JSON] [GradeInfo]
+
+    , eGetStudentGrades :: route
+        :- "students" :> Capture "studentAddr" Student
+        :> "grades"
+        :> Summary "Get all student's grades"
+        :> Description "Gets a list of all students grades (aka transactions \
+                        \in a private chain)"
+        :> Get '[JSON] [GradeInfo]
+
+    , eGetStudentCourseGrades :: route
+        :- "students" :> Capture "studentAddr" Student
+        :> "courses"  :> Capture "courseId" Course
+        :> "grades"
+        :> Summary "Get student's course grades"
+        :> Description "Gets a list of grades a student received during \
+                        \studying given course."
+        :> QueryParam "isFinal" IsFinal
+        :> Get '[JSON] [GradeInfo]
+
+      -- * Proofs
+
+    , eGetStudentProofs :: route
+        :- "students" :> Capture "studentAddr" Student
+        :> "proofs"
+        :> Summary "Get proofs of all student's activity"
+        :> Description "Gets all private transactions related to a student \
+                        \together with corresponding Merkle proofs."
+        :> Get '[JSON] [BlkProofInfo]
+
+    , eGetStudentCourseProofs :: route
+        :- "students" :> Capture "studentAddr" Student
+        :> "courses"  :> Capture "courseId" Course
+        :> "proofs"
+        :> Summary "Get proofs of student's course progress"
+        :> Description "Gets student's course grades and submissions in form \
+                        \of private transactions, together with corresponding \
+                        \Merkle proofs."
+        :> Get '[JSON] [BlkProofInfo]
+
+   } deriving (Generic)
```

Comment:
This looks nice; but is it possible to abstract "REST entity" of some kind?
I don't insist on doing this, though.

Reply:
What do you understand under `REST entity`?

Reply:
`POST /entity`
`GET /entity/`  -- get all
`GET /entity/:id`
`PUT /entity/:id`
`DELETE /entity/:id`

As one template or type in our case.

Reply:
Well, looks like each entity would have different set of endpoints corresponding to it (most have only all-getter, while many have no one-getter at all), so it's hard to extract such template.
I can visually group endpoints for each entity if you find that useful.